### PR TITLE
Merge `set_trial_state()` and `set_trial_values()` into one  function

### PR DESCRIPTION
**Pull request by masap**
## Motivation
In `tell()` of `Study` class, there is trailing code.

```
    if values is not None:
        self._storage.set_trial_values(trial_id, values)

    self._storage.set_trial_state(trial_id, state)
```

Consider process A is running ask-and-tell, process B is getting trial information by `study.get_trials()`. There are 3 steps at this code.

At step 1 of process A, process B get state=`TrialState.RUNNING`, value=`None`.
At step 2 of process A, process B get state=`TrialState.RUNNING`, value!=`None`.
At step 3 of process A, process B get state=`TrialState.COMPLETE`, value!=`None`.

```
    #step 1

    if values is not None:
        self._storage.set_trial_values(trial_id, values)

    #step 2

    self._storage.set_trial_state(trial_id, state)

    #step 3
```

We can get state=`TrialState.RUNNING`, value!=`None` at step 2 unexpectedly. It breaks consistency of trials because we expects only these combinations.
   - state=`TrialState.RUNNING`, value=`None`
   - state=`TrialState.COMPLETE`, value!=`None`

This issue is reported by @not522 in https://github.com/optuna/optuna/issues/2943#issuecomment-997744058.

## Description of the changes
To resolve the issue, this patch merges `set_trial_state()` and `set_trial_values()` into one function. To keep backward compatibility, we leaved `set_trial_state()` and `set_trial_values()` as deprecated.

### CODE REVIEW

**Comment by masap** [comment]
Some tests fail because http://azure.archive.ubuntu.com/ubuntu does not response. I will re-run later...

**Comment by masap** [comment]
Merge #3258 first, because of conflict. After #3258 is merged, I will fix the conflict.

**Comment by toshihikoyanase** [comment]
@keisuke-umezawa Could you examine this change is consistent with #2943 before diving into the details?

@not522 Thank you for your interest in this PR!

**Comment by masap** [comment]
Rebased.

**Comment by masap** [comment]
Rebased to fix conflict caused by #3258.

**Comment by keisuke-umezawa** [comment]
Before review, I would like to summarize the current spec of `set_trial_state_values`, 

1. `values` are set only when they are not `None` and trial is updatable.
2. `state` is set only when trial is updatable and state change is in the following pattern:
    - WAITING -> ANY
    - Others -> (COMPLETE, PRUNED, FAIL, WAITING)
3. The 1 and 2 are set in a atomic manner.

Also, the following things can be additional expected spec of `set_trial_state_values`, although we need more dicsussion.
1. `values` are set only when target state is `COMPLETE`.
2. etc.

**Comment by masap** [comment]
Fixed conflict.

**Comment by masap** [comment]
Rebased to retest.

File: optuna/storages/_cached_storage.py
Comment (keisuke-umezawa) [inline optuna/storages/_cached_storage.py:188]: Since we have not make `CachedStorage` and `InMemoryStorage` public, we can remove unnecessary methods. So, could you remove `set_trial_state()` and `set_trial_values()` from `CachedStorage` and `InMemoryStorage`?

File: optuna/storages/_in_memory.py
Comment (keisuke-umezawa) [inline optuna/storages/_in_memory.py:355]: Maybe, we can remove it.

File: optuna/storages/_in_memory.py
Comment (keisuke-umezawa) [inline optuna/storages/_in_memory.py:361]: If values are not None, we always set values.
After this line, we can write `self._set_trial(trial_id, trial)`?

**Review by keisuke-umezawa — CHANGES_REQUESTED** [review]
Add some comments. Could you check them?

File: optuna/storages/_base.py
Comment (masap) [reply optuna/storages/_base.py:461]: Thanks for your review!

> So, could you change it as abstractmethod and remove set_trial_state() and set_trial_values() from BaseStorage?
> Also, in this case, we do not need to provide default implementation for this method.

According to comment from @toshihikoyanase, there are users who has user-defined storage classes.

 - https://github.com/optuna/optuna/pull/3272#issuecomment-1025398438
 - https://github.com/optuna/optuna/pull/3272#issuecomment-1025472081

Visible storage classes for users are `RDBStorage` and `RedisStorage`. So I guess users inherit them to create user-defined storage classes.

If we remove `set_trial_state()` and `set_trial_values()` and make `set_trial_state_values()` returns `NotImplementedError`, it breaks backward compatibility.

How do you think about this ?

File: optuna/storages/_cached_storage.py
Comment (masap) [reply optuna/storages/_cached_storage.py]: Ok, fixed. I also remove remove `abstract` from `set_trial_state()` and `set_trial_values()` of `BaseStorage` class.

File: optuna/storages/_in_memory.py
Comment (masap) [reply optuna/storages/_in_memory.py]: Yes, it is duplicated. Removed.

File: optuna/storages/_in_memory.py
Comment (masap) [reply optuna/storages/_in_memory.py:310]: It makes code more simple, but when updating state we call `self._set_trial(trial_id, trial)` twice. So I selected more effective way. Which do you like ?

**Comment by masap** [comment]
Rebased for retest.

File: optuna/storages/_base.py
Comment (keisuke-umezawa) [reply optuna/storages/_base.py:461]: Thank you for the comment. We will discuss it and make decision for it.

Comment (keisuke-umezawa) [reply optuna/storages/_base.py:461]: cc: @toshihikoyanase 
After the discussion, we reached the following consensus.

Currently, as we have not make BaseStorage public yet, in our policy it should be allowed to add breaking changes into BaseStorage.

Could you follow the first comment by me?

> Since we have not make BaseStorage public, we can change which methods are abstract and remove unnecessary methods. So, could you change it as abstractmethod and remove set_trial_state() and set_trial_values() from BaseStorage?

File: optuna/storages/_in_memory.py
Comment (keisuke-umezawa) [reply optuna/storages/_in_memory.py:310]: I understand. Leave this line as it is.

File: optuna/storages/_base.py
Comment (masap) [reply optuna/storages/_base.py:417]: > Currently, as we have not make BaseStorage public yet, in our policy it should be allowed to add breaking changes into BaseStorage.

Ok, It is reasonable. I removed the methods and test for user defined storage.

**Review by keisuke-umezawa — APPROVED** [review]
LGTM! Thank you for working on this complicated issue.

@not522 Could you check this PR?

File: optuna/storages/_in_memory.py
Comment (not522) [inline optuna/storages/_in_memory.py:314]: When `state` is `RUNNING`, `values` should be `None`. These lines can be omitted.

File: optuna/storages/_rdb/storage.py
Comment (not522) [inline optuna/storages/_rdb/storage.py:650]: As default, we set `removed_version` as the two later major version of `deprecated_version`.
```suggestion
        "5.0.0",
```

File: optuna/storages/_rdb/storage.py
Comment (not522) [inline optuna/storages/_rdb/storage.py:775]:
```suggestion
        "5.0.0",
```

File: optuna/storages/_redis.py
Comment (not522) [inline optuna/storages/_redis.py:416]:
```suggestion
        "5.0.0",
```

File: optuna/storages/_redis.py
Comment (not522) [inline optuna/storages/_redis.py:552]:
```suggestion
        "5.0.0",
```

File: optuna/storages/_redis.py
Comment (not522) [inline optuna/storages/_redis.py:575]:
```suggestion
```

**Review by not522 — COMMENTED** [review]
Thank you for your great PR! I added minor comments. Please check them.
I think that we can remove the tests for `set_trial_state` and `set_trial_values`.

File: optuna/storages/_base.py
Comment (masap) [reply optuna/storages/_base.py]: Thank you for your review ! Fixed.

File: optuna/storages/_in_memory.py
Comment (masap) [reply optuna/storages/_in_memory.py]: Right ! Fixed. I also fixed related tests.

And I removed overwrite test from `test_set_trial_state_values_for_values` because I could not find out the way to update existing trial value. If it is possible please tell me.

```
        # Values can be overwritten.
        storage.set_trial_state_values(trial_id_1, state=trial_1.state, values=(0.2,))
        assert storage.get_trial(trial_id_1).value == 0.2
```

File: optuna/storages/_rdb/storage.py
Comment (masap) [reply optuna/storages/_rdb/storage.py]: Ok, Fixed.

File: optuna/storages/_rdb/storage.py
Comment (masap) [reply optuna/storages/_rdb/storage.py]: Fixed.

File: optuna/storages/_redis.py
Comment (masap) [reply optuna/storages/_redis.py]: Fixed.

File: optuna/storages/_redis.py
Comment (masap) [reply optuna/storages/_redis.py]: Fixed.

File: optuna/storages/_redis.py
Comment (masap) [reply optuna/storages/_redis.py]: Fixed.

**Comment by masap** [comment]
> I think that we can remove the tests for `set_trial_state` and `set_trial_values`.

How about remove the tests when the methods are removed ?

**Comment by not522** [comment]
> How about remove the tests when the methods are removed ?

OK, let's keep them as it is.

**Comment by not522** [comment]
Your PR looks good to me. But I realize my original idea is not smart -- `set_trial_state_values` requires complex if-statements. I will discuss the storages API design with other core developers this week. Please wait for other members' suggestions.

**Comment by masap** [comment]
> I will discuss the storages API design with other core developers this week. Please wait for other members' suggestions.

Ok.

**Review by not522 — APPROVED** [review]
Thank you for waiting. Let me merge this PR.
The storage API design is still under discussion. We may change it in follow-up PRs, but core developers agreed that it is not a blocker to merge this PR.

**Comment by masap** [comment]
Thanks!

**Pull request merged**
